src/OpenTelemetry.Instrumentation.EventCounters/EventCountersMetrics.cs

```diff
@@ -0,0 +1,100 @@
+
+namespace OpenTelemetry.Instrumentation.EventCounters;
+
+/// <summary>
+/// .NET EventCounters Instrumentation.
+/// </summary>
+internal class EventCountersMetrics : EventListener
+{
+    internal static readonly AssemblyName AssemblyName = typeof(EventCountersMetrics).Assembly.GetName();
+    internal static readonly Meter MeterInstance = new(AssemblyName.Name, AssemblyName.Version.ToString());
+
+    private readonly EventCountersInstrumentationOptions options;
+    private readonly ConcurrentBag<EventSource> preInitEventSources = new();
+    private readonly ConcurrentDictionary<Tuple<string, string>, Instrument> instruments = new();
+    private readonly ConcurrentDictionary<Tuple<string, string>, double> values = new();
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="EventCountersMetrics"/> class.
+    /// </summary>
+    /// <param name="options">The options to define the metrics.</param>
+    public EventCountersMetrics(EventCountersInstrumentationOptions options)
+    {
+        this.options = options;
+
+        while (this.preInitEventSources.TryTake(out EventSource source))
+        {
+            if (this.options.ShouldListenToSource(source.Name))
+            {
+                this.EnableEvents(source, EventLevel.LogAlways, EventKeywords.None, GetEnableEventsArguments(this.options));
+            }
+        }
+    }
+
+    /// <inheritdoc />
+    protected override void OnEventSourceCreated(EventSource source)
+    {
+        if (this.options == null)
+        {
+            this.preInitEventSources.Add(source);
+        }
+        else if (this.options.ShouldListenToSource(source.Name))
+        {
+            this.EnableEvents(source, EventLevel.LogAlways, EventKeywords.None, GetEnableEventsArguments(this.options));
+        }
+    }
+
+    /// <inheritdoc />
+    protected override void OnEventWritten(EventWrittenEventArgs eventData)
+    {
+        if (this.options == null)
+        {
+            return;
+        }
+
+        var payload = eventData.Payload[0] as IDictionary<string, object>;
+        var name = payload["Name"].ToString();
+        var isGauge = payload.ContainsKey("Mean");
+        Tuple<string, string> metricKey = new(eventData.EventSource.Name, name);
+
+        if (isGauge)
+        {
+            this.values[metricKey] = Convert.ToDouble(payload["Mean"]);
+        }
+        else
+        {
+            _ = this.values.TryAdd(metricKey, 0);
+            this.values[metricKey] += Convert.ToDouble(payload["Increment"]);
+        }
+
+        var instrumentName = $"{eventData.EventSource.Name}.{name}";
+        _ = this.instruments.TryAdd(
+            metricKey,
+            isGauge ? MeterInstance.CreateObservableGauge(instrumentName, () => this.values[metricKey]) : MeterInstance.CreateObservableCounter(instrumentName, () => this.values[metricKey]));
+    }
+
+    private static Dictionary<string, string> GetEnableEventsArguments(EventCountersInstrumentationOptions options) =>
```

Comment:
Return type need to be changed to `Dictionary<string, string?>`

Reply:
Why, `EnableEvents` takes an `IDictionary<string, string>` as the `arguments` parameter?